### DISCUSSION TITLE
Can we copy every profile of project inside apparmor.d directory?

### DISCUSSION THREAD

**Discussion by osevan** [roddhjav/apparmor.d#213] [General]
Can we do this on already running apparmor protected system?

Im asking because why go pkg and recompile of apparmor lsm is needed and additonal install notes required. 

Thanks and

Best regards

**Comment by roddhjav** [comment]
**Short:** There is a reason why it works this way.

I am not sure I understand your question:

* > Can we do this on already running apparmor protected system?

  Yes, having apparmor enabled and installed is actually a requirement as this project is only providing additional profiles.

* > Im asking because why go pkg and recompile of apparmor lsm is needed and additonal install notes required.

  There is no additional install setup. The only thing you have to do is to build and install a package (tailored for your distribution).

**Long:** When you run `make` the following is done on the profiles:
- Merge all profiles in one unified directory as apparmor is not compatible with the `groups` and `profile-x-y` directory structure.
- Bypass userspace tools restriction (See #76 and #87). This is a long running bug in upstream that prevents the use of variables in profile attachment for userspace tools (aa-complain, aa-logprof...). The [fix](https://github.com/roddhjav/apparmor.d/blob/main/pkg/aa/variables_test.go) in apparmor.d is the reason why this prebuild is in go and not in bash (it used to be in bash).
- Set complain/enforce mode on all profiles
- Apply distribution specific settings: extra abstraction, profiles, fix incompatibilities
- Ignore profile ignored in `dists/ignore/` 
- Set flags configuration from `dists/ignore/`
- Some drop-in system config file are needed to ensure apparmor profile are loaded before some services
- Compile `aa-log`

**Comment by osevan** [comment]
Could you go deeper here?

"Some drop-in system config file are needed to ensure apparmor profile are loaded before some services"

Its possible to load apparmor as pid1?

I mean before init system?

**Comment by roddhjav** [comment]
We simply need to ensure the profiles are loaded into the kernel before the start of some services are loaded (see the [systemd](https://github.com/roddhjav/apparmor.d/tree/main/systemd/system) directory). It simply means we need to ensure that systemd is going to start apparmor.service *before* services such as `systemd-journald.service`, `systemd-networkd.service`...

Note: there are [other](https://gitlab.com/apparmor/apparmor/-/wikis/AppArmorInSystemd#early-policy-loads) [methods](https://gitlab.com/apparmor/apparmor/-/wikis/FullSystemPolicy#implementation) to actually load the profiles into the kernel before even starting systemd, however, it is not used (yet) in this project.

**Comment by osevan** [comment]
Userspace tools like aa-complain .*. Still works inside apparmor.d directory ?

aa-enforce  .*. Too?
aa-logorof too like everytime? 


After installed it is in enforcement or in complain mode by default?

**Comment by roddhjav** [comment]
All aa tools work as expected. As explained in the documentation, the profile are in complain mode by default, then you can put then in enforce mode: https://apparmor.pujol.io/enforce/

**Comment by osevan** [comment]
Udisk makes trouble,  when usb stick plugged in

Os= artix linux.

Every profile are in complain mode

Only works again when sudo aa-teardown and module is unloaded.

**Reply by roddhjav** [reply]
Please share the apparmor log, so I can fix the issue. See https://apparmor.pujol.io/report/

Also please create an issue for this, a discussion is not the best place for bug report.

**Comment by osevan** [comment]
sudo aa-disable dbus-daemon-launch-helper

Helped. 

But no log output..
Where should i look?

**Reply by osevan** [reply]
Jup i double checked no logs spit out, but disabling this above helped.
